Zachmu/bug fixes to contribute

zachmu opened the pull request:
Several bug fixes and improvements:

1) Float comparisons were broken because they were being truncated to integers. Fixed and added many tests with float values.
2) Negative float values were getting optimized away to 0 by the parser, fixed and added tests.
3) Added integration tests of IS NULL, IS NOT NULL
4) Implemented IS TRUE, IS FALSE, and NOT versions of each
5) Fixed some typos in the source
6) Allow conversion of string > bool to match MySQL semantics
7) Fixed bug where some non-column filter expressions were getting effectively optimized away by being force-converted to a boolean

zachmu commented:
Sorry for the relatively messy diff (includes several recent changes from your master), not sure what caused it exactly. Open to suggestions for how to make it cleaner.

zachmu commented:
Cleaned this up with the help of my colleague, ready to be reviewed.

erizocosmico commented on sql/expression/istrue.go:
```diff
@@ -0,0 +1,77 @@
+	}
+	if e.invert {
+		return f(NewIsFalse(child))
+	} else {
```
You can just remove the else branch entirely, since the if branch returns

zachmu replied:
Done.

juanjux commented on sql/expression/istrue.go:
```diff
@@ -0,0 +1,77 @@
+
+import "github.com/src-d/go-mysql-server/sql"
+
+// IsNull is an expression that checks if an expression is true.
```
s/IsNull/IsTrue

zachmu replied:
Done.

juanjux commented on sql/expression/istrue_test.go:
```diff
@@ -0,0 +1,49 @@
+	require.Equal(false, eval(t, e, sql.NewRow(nil)))
+	require.Equal(true, eval(t, e, sql.NewRow(100)))
+	require.Equal(true, eval(t, e, sql.NewRow(-1)))
+	require.Equal(false, eval(t, e, sql.NewRow(0)))
```
Could you add some tests with other data types (like strings)?

zachmu replied:
Done. Also caught a bug related to nil values :D

juanjux commented:
Only a couple easy to fix things, thanks for the PR!

ajnavarro commented on sql/analyzer/optimization_rules.go:
```diff
@@ -389,6 +389,16 @@ func evalFilter(ctx *sql.Context, a *Analyzer, node sql.Node) (sql.Node, error)
+				// UnaryMinus expressions come back from the parser when a negative float is evaluated. Treat them just like
+				// normal literal expressions.
+				if um, ok := e.(*expression.UnaryMinus); ok {
+					negated, err := e.Eval(ctx, nil)
```
Can we have here the case that UnaryMinus is wrapping a non literal child?

zachmu replied:
That should be covered by the isEvaluable check above (returns false if an expression contains columns). I don't want to assert a Literal child here because it wouldn't work for weird arithmetic expressions e.g. -(60*60*24).

zachmu commented:
Addressed PR feedback and updated to include the recent interface change to Expression.

Adding tests for boolean columns for IsTrue evaluation as Juanjo suggested required being able to evaluate strings as booleans. Once implemented, this exposed a bug in the filter optimizer, where non-column expressions were incorrectly being coerced to booleans. Really, this is the same bug that was causing UnaryMinus expressions (for negative floats) to not work correctly. My fix is more general than my original PR: don't coerce to boolean during optimization. This is slightly less performant but more correct.

Please take another look and merge if it looks good.

juanjux commented:
Thanks for the fixes!
